src/library/export/engineprimeexportjob.cpp

```diff
@@ -0,0 +1,498 @@
+}
+
+void EnginePrimeExportJob::trackLoaded(TrackRef trackRef, TrackPointer trackPtr) {
+    // See if this track is in our queue.
+    auto index = m_trackRefs.indexOf(trackRef);
+    if (index == -1) {
+        // Not a track we're interested in.
+        return;
+    }
+
+    // Enqueue the loaded track onto the queue, and notify that it is loaded.
+    QMutexLocker lock{&m_trackMutex};
+    m_loadedTrackQueue.enqueue(trackPtr);
+    m_tracksLoaded++;
+    m_waitAnyTrack.wakeAll();
+}
+
+// Obtain a set of all track refs across all directories in the whole Mixxx library.
+QSet<TrackRef> EnginePrimeExportJob::getAllTrackRefs() const {
+    QSet<TrackRef> trackRefs;
+    auto dirs = m_trackCollectionManager.internalCollection()
+                        ->getDirectoryDAO()
+                        .getDirs();
+    for (auto iter = dirs.cbegin(); iter != dirs.cend(); ++iter) {
+        trackRefs.unite(m_trackCollectionManager.internalCollection()
+                                ->getTrackDAO()
+                                .getAllTrackRefs(*iter)
+                                .toSet());
+    }
+    return trackRefs;
+}
+
+// Obtain a set of track refs in a set of crates.
+QSet<TrackRef> EnginePrimeExportJob::getTracksRefsInCrates(
+        const QSet<CrateId>& crateIds) const {
+    QSet<TrackRef> trackRefs;
+    for (auto iter = crateIds.cbegin(); iter != crateIds.cend(); ++iter) {
+        auto result = m_trackCollectionManager.internalCollection()
+                              ->crates()
+                              .selectCrateTracksSorted(*iter);
+        while (result.next()) {
+            auto trackId = result.trackId();
+            auto location = m_trackCollectionManager.internalCollection()
+                                    ->getTrackDAO()
+                                    .getTrackLocation(trackId);
+            auto trackFile = TrackFile(location);
+            trackRefs.insert(TrackRef::fromFileInfo(trackFile, trackId));
+        }
+    }
+    return trackRefs;
+}
+
+void EnginePrimeExportJob::run() {
+    // Crate music directory if it doesn't already exist.
+    QDir().mkpath(m_request.musicFilesDir.path());
+
+    // Determine how many tracks and crates we have to export, and use to
+    // calculate job progress.
+    QSet<TrackRef> trackRefs;
+    QSet<CrateId> crateIds;
+    if (m_request.exportSelectedCrates) {
+        trackRefs = getTracksRefsInCrates(m_request.crateIdsToExport);
+        crateIds = m_request.crateIdsToExport;
+    } else {
+        // Note that we do not currently export empty crates.
+        trackRefs = getAllTrackRefs();
+
+        // Make a list of track ids, and get crates that contain those tracks.
+        QList<TrackId> trackIds;
+        for (auto iter = trackRefs.cbegin(); iter != trackRefs.cend(); ++iter) {
+            trackIds.append(iter->getId());
+        }
+        crateIds = m_trackCollectionManager.internalCollection()
+                           ->crates()
```

Comment:
Using a QObject like `TrackCollectionManager` in another thread is not allowed and causes undefined behaviour!!

Reply:
@uklotzde What do you propose how we should read information about the track collection? Can this only be done from the GUI thread? If yes, then it's weird to have such a core functionality of the Mixxx “engine” coupled to the GUI.

Reply:
> Can this only be done from the GUI thread?

In the absence of a better way, this was the refactoring that I was planning to make to fix this issue.

Reply:
@haslersn This is a known architectural deficiency of Mixxx. Fixing it would require rewriting major parts of Mixxx because there is so much coupling and almost no isolation. Ideas (and contributions) that would allow us to gradually overcome these limitations are very welcome.

Reply:
Ok, that's what I thought. My question “What do you propose [...]” was regarding the fashion in which we offload the work to the GUI thread. My idea would be

* either to read all of the library metadata at the beginning of the export (This might lead to a short freeze of the GUI)
* or to request metadata in small chunks and handle these requeste from the GUI thread. Is there such a request/response mechanism in Qt?

Reply:
After talking to @haslersn yesterday for a bit, we (I assume he is with me here) think that the first option would be the better one, as the implementation would be less complex. Also, a short freeze of the GUI wouldn't be _too_ bad, as we (for now) would not expect the user to do something in the GUI during the export.

Later, this could easily be refactored to a future that collects the metadata in the GUI thread and as soon as it is ready, the actual export starts, unblocking the GUI thread this way.

I did not look at the code for this idea (because I don't even know where it is) and am assuming that this is a possible way of refactoring it.

Reply:
Start with the simple implementation that just works.

We need a common, generic pattern on how to solve the async request/response issues within Mixxx using Qt's signals/slots. There are already some attempts in `library/TrackLoader`, `network/WebTask`, and in #2282, but nothing that could be reused easily. From my experience so far it requires a lot of boilerplate code :P

Reply:
The UI blocking issues can be mitigated by using a modal progress dialog that allows to abort long running tasks.

Reply:
Hi @uklotzde - @haslersn has kindly drafted a commit which uses a thread-local copy of `CrateStorage` and `AnalysisDao`, so that the interaction with these classes can safely happen on the worker thread.  Before I merge this commit in, do you have a strong opinion on this approach vs. loading on the GUI thread?  Thanks.

Reply:
The commit can be found here: https://github.com/haslersn/mixxx/commit/833816a38595f4e29007c1621b9be1547620ae19

Reply:
Those classes are safe to use with a thread-local database connection. Especially CrateStorage which is a dumb facade that doesn't even inherit from QObject.

Nevertheless accessing the same database from different threads might cause locks and failed queries, especially when executing write operations and using long running queries like the LibraryScanner does (https://bugs.launchpad.net/mixxx/+bug/1051106).

Since you are only performing read operations (probably without a transaction) you may give it a try.

Reply:
For various reasons, I ended up not using the thread-local copies commit.  Instead, I've now pushed an update that moves all interaction with `TrackCollectionManager` onto the GUI thread and has various assertions to check the thread affinity.

As we now have a consistent approach to off-loading to the GUI thread, I was also able to remove use of `TrackLoader` and instead go to the `TrackCollection` directly.

I've tested this on Ubuntu Xenial and Fedora 32, and both work well.

Reply:
I just recently discovered that synchronous invocations using Qt::BlockingQueuedConnection might be sufficient for many cases. But in order to use this pattern we need to split our interfaces into thread-safe public methods that forward to internal slots, either directly within the same thread or using a queued invocation from different threads.